presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -165,38 +167,62 @@ public static PlanNodeStatsEstimate addStatsAndCollapseDistinctValues(PlanNodeSt
+        double estimatedRowCount = Math.min(left.getOutputRowCount(), right.getOutputRowCount());
+
+        List<Double> allRowCounts = buildVariableStatistics(
+                left, right, statsBuilder, estimatedRowCount, StatisticRange::intersect);
+
+        double preciseRowCount = allRowCounts.stream().reduce(estimatedRowCount, Double::min);
+        return statsBuilder.setOutputRowCount(preciseRowCount).build();
+    }
+
+    private static PlanNodeStatsEstimate addStats(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right,
+            RangeAdditionStrategy rangeAdder)
     {
         if (left.isOutputRowCountUnknown() || right.isOutputRowCountUnknown()) {
             return PlanNodeStatsEstimate.unknown();
         }
 
         PlanNodeStatsEstimate.Builder statsBuilder = PlanNodeStatsEstimate.builder();
-        double newRowCount = left.getOutputRowCount() + right.getOutputRowCount();
+        double rowCount = left.getOutputRowCount() + right.getOutputRowCount();
+        buildVariableStatistics(left, right, statsBuilder, rowCount, rangeAdder);
+
+        return statsBuilder.setOutputRowCount(rowCount).build();
+    }
 
+    private static List<Double> buildVariableStatistics(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right,
+            PlanNodeStatsEstimate.Builder statsBuilder, double estimatedRowCount, RangeAdditionStrategy rangeAdder)
+    {
+        List<Double> allRowCounts = new LinkedList<>();
         concat(left.getVariablesWithKnownStatistics().stream(), right.getVariablesWithKnownStatistics().stream())
                 .distinct()
                 .forEach(symbol -> {
                     VariableStatsEstimate symbolStats = VariableStatsEstimate.zero();
-                    if (newRowCount > 0) {
+                    if (estimatedRowCount > 0) {
                         symbolStats = addColumnStats(
                                 left.getVariableStatistics(symbol),
                                 left.getOutputRowCount(),
                                 right.getVariableStatistics(symbol),
                                 right.getOutputRowCount(),
-                                newRowCount,
-                                strategy);
+                                estimatedRowCount,
+                                rangeAdder);
                     }
                     statsBuilder.addVariableStatistics(symbol, symbolStats);
+                    allRowCounts.add(symbolStats.getDistinctValuesCount());
                 });
-
-        return statsBuilder.setOutputRowCount(newRowCount).build();
+        return allRowCounts;
     }
 
-    private static VariableStatsEstimate addColumnStats(VariableStatsEstimate leftStats, double leftRows, VariableStatsEstimate rightStats, double rightRows, double newRowCount, RangeAdditionStrategy strategy)
+    private static VariableStatsEstimate addColumnStats(VariableStatsEstimate leftStats, double leftRows,
```

Comment:
why did you get rid of the argument check? also, same comment about formatting method arguments.

Reply:
This was indeed a mistake while moving the lines of code around. Thanks for catching it. I will bring this back.